File: MaquinaMoore.java
import java.io.IOException;

public class MaquinaMoore extends AnalisadorLexico {

    public MaquinaMoore(String nome) {
        super(nome);
    }

    public void s0() {
        if (this.proximoCaractereIs(ABREPAR)) {
            leProxCaractere();
            s1();
        }

        else if (this.proximoCaractereIs(ABRECOL)) {
            leProxCaractere();
            s2();
        }

        else if (this.proximoCaractereIs(ABRECHV)) {
            leProxCaractere();
            s3();
        }

        else if (this.proximoCaractereIs(FECHAPAR)) {
            leProxCaractere();
            s4();
        }

        else if (this.proximoCaractereIs(FECHACOL)) {
            leProxCaractere();
            s5();
        }

        else if (this.proximoCaractereIs(FECHACHV)) {
            leProxCaractere();
            s6();
        }

        else if (this.proximoCaractereIs(NUM)) {
            leProxCaractere();
            s7();
        }

        else if (this.proximoCaractereIs(LETRAS)) {
            leProxCaractere();
            s8();
        }

        else if (this.proximoCaractereIs(OP)) {
            leProxCaractere();
            s9();
        }

        else if (this.proximoCaractereIs(EOF)) {
            leProxCaractere();
            s10();
        }

        else {
            throw new ErroLexico(this.proximoCaractere, NUM + RESTO_VAR + OP + ABRECHV + ABRECOL + ABREPAR + FECHACHV + FECHACOL + FECHAPAR);
        }
    }

    public void s1() {
        this.tokenReconhecido = Token.ABREPAR;
    }

    public void s2() {
        this.tokenReconhecido = Token.ABRECOL;
    }

    public void s3() {
        this.tokenReconhecido = Token.ABRECHV;
    }

    public void s4() {
        this.tokenReconhecido = Token.FECHAPAR;
    }

    public void s5() {
        this.tokenReconhecido = Token.FECHACOL;
    }

    public void s6() {
        this.tokenReconhecido = Token.FECHACHV;
    }

    public void s7() {
        this.tokenReconhecido = Token.NUM;

        if (this.proximoCaractereIs(NUM)) {
            leProxCaractere();
            s7();
        }        
    }
    
    public void s8() {
        this.tokenReconhecido = Token.VAR;

        if (this.proximoCaractereIs(RESTO_VAR)) {
            leProxCaractere();
            s8();
        }
    }
    
    public void s9() {
        this.tokenReconhecido = Token.OP;
    }
    
    public void s10() {
        this.tokenReconhecido = Token.EOF;
    }
}

File: analisadorLexicoC++/MaquinaMoore.cpp
#include "MaquinaMoore.h"
#include "ErroLexico.h"

MaquinaMoore::MaquinaMoore(const std::string& nome)
    : AnalisadorLexico(nome)
{
}

void MaquinaMoore::s0() {

    while (proximoCaractereIs(Constantes::SEPARADOR)) {
        leProxCaractere();
    }

    if (proximoCaractereIs(Constantes::ABREPAR)) {
        leProxCaractere();
        s1();
    }

    else if (proximoCaractereIs(Constantes::ABRECOL)) {
        leProxCaractere();
        s2();
    }

    else if (proximoCaractereIs(Constantes::ABRECHV)) {
        leProxCaractere();
        s3();
    }

    else if (proximoCaractereIs(Constantes::FECHAPAR)) {
        leProxCaractere();
        s4();
    }

    else if (proximoCaractereIs(Constantes::FECHACOL)) {
        leProxCaractere();
        s5();
    }

    else if (proximoCaractereIs(Constantes::FECHACHV)) {
        leProxCaractere();
        s6();
    }

    else if (proximoCaractereIs(Constantes::NUM)) {
        leProxCaractere();
        s7();
    }

    else if (proximoCaractereIs(Constantes::LETRAS)) {
        leProxCaractere();
        s8();
    }

    else if (proximoCaractere == Constantes::EOF_CHAR) {
        leProxCaractere();
        s10();
    }

    else if (proximoCaractereIs(Constantes::OP)) {
        leProxCaractere();
        s9();
    }

    else {
        throw ErroLexico(
            proximoCaractere,
            Constantes::NUM
            + Constantes::RESTO_VAR
            + Constantes::OP
            + Constantes::ABRECHV
            + Constantes::ABRECOL
            + Constantes::ABREPAR
            + Constantes::FECHACHV
            + Constantes::FECHACOL
            + Constantes::FECHAPAR
        );
    }
}

void MaquinaMoore::s1() {
    tokenReconhecido = Constantes::Token::ABREPAR;
}

void MaquinaMoore::s2() {
    tokenReconhecido = Constantes::Token::ABRECOL;
}

void MaquinaMoore::s3() {
    tokenReconhecido = Constantes::Token::ABRECHV;
}

void MaquinaMoore::s4() {
    tokenReconhecido = Constantes::Token::FECHAPAR;
}

void MaquinaMoore::s5() {
    tokenReconhecido = Constantes::Token::FECHACOL;
}

void MaquinaMoore::s6() {
    tokenReconhecido = Constantes::Token::FECHACHV;
}

void MaquinaMoore::s7() {

    tokenReconhecido = Constantes::Token::NUM;

    if (proximoCaractereIs(Constantes::NUM)) {

        leProxCaractere();
        s7();
    }

    else if (!(proximoCaractereIs(Constantes::SEPARADOR)
            || proximoCaractere == Constantes::EOF_CHAR
            || proximoCaractereIs(Constantes::ABREPAR)
            || proximoCaractereIs(Constantes::ABRECHV)
            || proximoCaractereIs(Constantes::ABRECOL)
            || proximoCaractereIs(Constantes::FECHAPAR)
            || proximoCaractereIs(Constantes::FECHACOL)
            || proximoCaractereIs(Constantes::FECHACHV)
            || proximoCaractereIs(Constantes::OP))) {

        throw ErroLexico(
            proximoCaractere,
            "separador ou fim de Token"
        );
    }
}

void MaquinaMoore::s8() {

    tokenReconhecido = Constantes::Token::VAR;

    if (proximoCaractereIs(Constantes::RESTO_VAR)) {

        leProxCaractere();
        s8();
    }

    else if (!(proximoCaractereIs(Constantes::SEPARADOR)
            || proximoCaractere == Constantes::EOF_CHAR
            || proximoCaractereIs(Constantes::ABREPAR)
            || proximoCaractereIs(Constantes::ABRECHV)
            || proximoCaractereIs(Constantes::ABRECOL)
            || proximoCaractereIs(Constantes::FECHAPAR)
            || proximoCaractereIs(Constantes::FECHACOL)
            || proximoCaractereIs(Constantes::FECHACHV)
            || proximoCaractereIs(Constantes::OP))) {

        throw ErroLexico(
            proximoCaractere,
            "separador ou fim de Token"
        );
    }
}

void MaquinaMoore::s9() {
    tokenReconhecido = Constantes::Token::OP;
}

void MaquinaMoore::s10() {
    tokenReconhecido = Constantes::Token::EOF_TOKEN;
}